pkg/ccl/backupccl/restore_old_versions_test.go

```diff
@@ -248,3 +250,194 @@ func restoreOldVersionClusterTest(exportDir string) func(t *testing.T) {
+			checksums = append(checksums, checksumBankPayload(t, sqlDB))
+
+			backupDir := fmt.Sprintf("nodelocal://0/%d", backupNum)
+			var from string
+			if backupNum > 0 {
+				from = fmt.Sprintf(` INCREMENTAL FROM %s`, strings.Join(backupDirs, `,`))
+			}
+			sqlDB.Exec(t, fmt.Sprintf(`BACKUP TO '%s' %s`, backupDir, from))
+			blockBackfill <- struct{}{}
+			require.NoError(t, <-createErr)
+
+			backupDirs = append(backupDirs, fmt.Sprintf(`'%s'`, backupDir))
+		}
+
+		// Test a regression in RESTORE where the batch end key was not
+		// being set correctly in Import: make an incremental backup such that
+		// the greatest key in the diff is less than the previous backups.
+		sqlDB.Exec(t, `INSERT INTO data.bank VALUES (0, -1, 'final')`)
+		checksums = append(checksums, checksumBankPayload(t, sqlDB))
+		sqlDB.Exec(t, fmt.Sprintf(`BACKUP TO '%s' %s`,
+			"nodelocal://0/final", fmt.Sprintf(` INCREMENTAL FROM %s`, strings.Join(backupDirs, `,`)),
+		))
+		backupDirs = append(backupDirs, `'nodelocal://0/final'`)
+	}
+	os.Rename(dir, path/to/testdata)
+*/
+
+// TestRestoreOldBackupMissingOfflineIndexes tests restoring a backup made by
+// v20.2 prior to the introduction of excluding offline indexes in #62572 using
+// the commented-out code above in TestCreateIncBackupMissingIndexEntries. Note:
+// that code needs to be pasted into a branch checkout _prior_ to the inclusion
+// of the mentioned PR.
+func TestRestoreOldBackupMissingOfflineIndexes(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	ctx := context.Background()
+
+	badBackups, err := filepath.Abs("testdata/restore_old_versions/inc_missing_addsst/v20.2.7")
+	require.NoError(t, err)
+	args := base.TestServerArgs{ExternalIODir: badBackups}
+	backupDirs := make([]string, 9)
+	for i := range backupDirs {
+		backupDirs[i] = fmt.Sprintf("'nodelocal://0/%d'", i)
+	}
+
+	// Start a new cluster to restore into.
+	{
+		for i := len(backupDirs); i > 0; i-- {
+			restoreTC := testcluster.StartTestCluster(t, singleNode, base.TestClusterArgs{ServerArgs: args})
+			defer restoreTC.Stopper().Stop(context.Background())
+			sqlDBRestore := sqlutils.MakeSQLRunner(restoreTC.Conns[0])
+			from := strings.Join(backupDirs[:i], `,`)
+			sqlDBRestore.Exec(t, fmt.Sprintf(`RESTORE FROM %s`, from))
+
+			for j := i; j > 1; j-- {
+				var res int64
+				switch j % 3 {
+				case 2:
+					for i := 0; i < 50; i++ {
+						if err := sqlDBRestore.DB.QueryRowContext(ctx,
+							fmt.Sprintf(`SELECT count(*) FROM data.bank@test_idx_%d`, j-1),
+						).Scan(&res); err != nil {
+							if !strings.Contains(err.Error(), `not found`) {
+								t.Fatal(err)
+							}
+							t.Logf("index %d doesn't exist yet on attempt %d", j-1, i)
+							time.Sleep(time.Millisecond * 50)
+							continue
+						}
+						break
+					}
+					var expected int64
+					sqlDBRestore.QueryRow(t, `SELECT count(*) FROM data.bank@primary`).Scan(&expected)
+					if res != expected {
+						t.Fatalf("got %d, expected %d", res, expected)
+					}
+				case 0:
```

Comment:
Was the exclusion of `case 1`/geo indexes intentional? I don't think I follow why those are different.

Reply:
Added a comment